devito/dse/manipulation.py

```diff
@@ -149,36 +150,34 @@ def common_subexprs_elimination(exprs, make, mode='default'):
         # Apply replacements
         processed = [e.xreplace(mapper) for e in processed]
         mapped = [e.xreplace(mapper) for e in mapped]
-        mapped = [Eq(v, k) for k, v in reversed(list(mapper.items()))] + mapped
+        mapped = [DummyEq(v, k) for k, v in reversed(list(mapper.items()))] + mapped
 
         # Prepare for the next round
         for k in picked:
             targets.pop(k)
     processed = mapped + processed
 
+    # At this point we may have useless temporaries (e.g., r0=r1). Let's drop them
+    processed = compact_temporaries(processed)
+
     # Perform topological sorting so that reads-after-writes are honored
     processed = topological_sort(processed)
 
     return processed
 
 
-def compact_temporaries(temporaries, leaves):
+def compact_temporaries(exprs):
     """Drop temporaries consisting of isolated symbols."""
-    exprs = temporaries + leaves
-    targets = {i.lhs for i in leaves}
-
     graph = FlowGraph(exprs)
 
     mapper = {k: v.rhs for k, v in graph.items()
-              if v.is_Scalar and
-              (q_leaf(v.rhs) or v.rhs.is_Function) and
-              not v.readby.issubset(targets)}
+              if v.is_Scalar and (q_leaf(v.rhs) or v.rhs.is_Function)}
 
     processed = []
     for k, v in graph.items():
         if k not in mapper:
             # The temporary /v/ is retained, and substitutions may be applied
-            handle, _ = xreplace_constrained(v, mapper, repeat=True)
+            handle, _ = yreplace(v, mapper, repeat=True)
```

Comment:
why "y"replace?

Reply:
xreplace_constrained was a horrible name. Long and horrible

`yreplace` reminds me of something smarter than `xreplace`. `y` follows `x` alphabetically...if you have a fancier name please shoot

Reply:
I don't know the name doesn't feel very explicit or explanatory. Maybe something like `custom_replace` or `deep_replace`